transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
 func (t Transaction) validateChecksum() error {
 	// Check if finalPage is set
-	if t.finalPage == nil {
-		return errors.New("Couldn't verify checksum. finalPage is nil")
+	if t.firstPage == nil {
+		return errors.New("Couldn't verify checksum. firstPage is nil")
 	}
 	checksum, err := t.checksum()
 	if err != nil {
 		return errors.New("Failed to create checksum for validation")
 	}
-	if checksum != t.finalPage.transactionChecksum {
+	if checksum != t.firstPage.transactionChecksum {
 		return errors.New("checksum not valid")
 	}
 	return nil
 }
 
-//commit commits a transaction by setting the correct status and checksum
-func (t *Transaction) commit() error {
-	// set the status of the first page first
-	t.firstPage.pageStatus = pageStatusComitted
-
-	// calculate the checksum and write it to the last page
-	checksum, err := t.checksum()
-	if err != nil {
-		return build.ExtendErr("Unable to create checksum of transaction", err)
-	}
-	t.finalPage.transactionChecksum = checksum
-	err = t.finalPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the final page failed", err)
+// NewTransaction creates a transaction from a set of updates
+func (w *WAL) NewTransaction(updates []Update) *Transaction {
+	// Create new transaction
+	newTransaction := Transaction{
+		Updates:      updates,
+		wal:          w,
+		initComplete: make(chan error),
 	}
 
-	// Finalize the commit by writing the first page with the updated status
-	err = t.firstPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the first page failed", err)
-	}
+	// Initialize the transaction by splitting up the payload among free pages
+	// and writing them to disk.
+	go threadedInitTransaction(&newTransaction)
 
-	t.commitComplete = true
-	t.wal.logFile.Sync()
-	return nil
+	return &newTransaction
 }
 
-// SignalApplyComplete  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
-func (t *Transaction) SignalApplyComplete() <-chan error {
+// SignalUpdatesApplied  informs the WAL that it is safe to free the used pages to reuse them in a new transaction
+func (t *Transaction) SignalUpdatesApplied() <-chan error {
 	if !t.setupComplete || !t.commitComplete || t.releaseComplete {
-		panic("misuse of transaction - call each of the signaling methods exactly ones, in serial, in order")
+		panic("misuse of transaction - call each of the signaling methods exactly once, in serial, in order")
 	}
 	t.releaseComplete = true
 	notifyChannel := make(chan error)
 
 	go func() {
-		t.wal.mu.Lock()
-		defer t.wal.mu.Unlock()
-
 		// Set the page status to applied
 		t.firstPage.pageStatus = pageStatusApplied
 
 		// Write the page to disk
-		err := t.firstPage.Write(t.wal.logFile)
+		var err error
+		if t.wal.deps.disrupt("ReleaseFail") {
+			// Disk failure causes the commit to fail
+			err = errors.New("Write failed on purpose")
+		} else {
+			err = t.firstPage.Write(t.wal.logFile)
```

Comment:
If we end up changing the design so that WriteToFile writes all the pages, then we probably would want a second helper function `Release` or something that writes just the single updated field to disk. Not fully sure though how best  to do that.